mlora/MixLoRA.py

```diff
@@ -0,0 +1,133 @@
+from mlora.modelargs import MultiLoraBatchData
+from mlora.LoraLiner import Linear
+
+import torch
+import torch.nn.functional as F
+from typing import Dict, Tuple, Optional
+
+
+class MixGate(torch.nn.Module):
+    def __init__(self, adapter_name: str) -> None:
+        super().__init__()
+
+        self.adapter_name_: str = adapter_name
+        self.gate_: torch.nn.Linear = None
+        self.experts_: int = 8
+        self.topk_: int = 2
+
+    def set_parameter(self, moe_in_features: int, moe_experts: int, moe_topk: int, device: str):
+        self.gate_ = torch.nn.Linear(
+            moe_in_features,
+            moe_experts,
+            bias=False,
+            device=device,
+        )
+        self.experts_ = moe_experts
+        self.topk_ = moe_topk
+
+    def routing(self, norm_data: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+        # routing to experts
+        router_logits = self.gate_.forward(norm_data)
+        routing_weights = F.softmax(
+            router_logits, dim=1, dtype=torch.float)
+        routing_weights, selected_experts = torch.topk(
+            routing_weights, self.topk_, dim=-1
+        )
+        routing_weights /= routing_weights.sum(dim=-1, keepdim=True)
+        return routing_weights, selected_experts
+
+    def forward(self, routing_state: Tuple[torch.Tensor, torch.Tensor],
+                hidden_state: torch.Tensor, expert_idx: int) -> torch.Tensor:
+        routing_weights, selected_experts = routing_state
+        expert_mask = selected_experts == expert_idx
+        expert_weights = (routing_weights * expert_mask).sum(
+            dim=-1, keepdim=True
+        )
+        return hidden_state.mul_(expert_weights)
+
+
+class MixMoe(torch.nn.Module):
+    def __init__(self, w1: Linear, w2: Linear, w3: Linear) -> None:
+        super().__init__()
+
+        # feed forward
+        self.w1_: Linear = w1  # also gate FNN * dim
+        self.w2_: Linear = w2  # also down dim * FNN
+        self.w3_: Linear = w3  # also up   FNN * dim
+        # mix of experts
+        self.moes_: Dict[str, MixGate] = {}
+
+    def init_moe_weight(self, adapter_name: str,
+                        moe_in_features: int,
+                        moe_experts: int,
+                        moe_topk: int,
+                        gate: Optional[torch.Tensor] = None):
+        if adapter_name not in self.moes_:
+            self.moes_[adapter_name] = MixGate(adapter_name)
+
+        self.moes_[adapter_name].set_parameter(
+            moe_in_features, moe_experts, moe_topk, self.w1_.device_)
+
+        if gate is not None:
+            with torch.no_grad():
+                self.moes_[adapter_name].gate_.weight.copy_(gate)
+
+    def forward(self, score_norm_data: torch.Tensor, input_args: MultiLoraBatchData) -> torch.Tensor:
+        common_w1 = self.w1_.weight_.forward(score_norm_data)
+        common_w3 = self.w3_.weight_.forward(score_norm_data)
```

Comment:
where is the common_w2? any meaning here?

Reply:
The calculation of MLP is w1, w3 -> w2, so among multiple experts, the shared w1 w3 can be calculated first to save calculation overhead.